cmake/OpenCVDownload.cmake

```diff
@@ -37,6 +37,47 @@ file(WRITE "${OPENCV_DOWNLOAD_LOG}" "#use_cache \"${OPENCV_DOWNLOAD_PATH}\"\n")
 file(REMOVE "${OPENCV_DOWNLOAD_WITH_CURL}")
 file(REMOVE "${OPENCV_DOWNLOAD_WITH_WGET}")
 
+# host
+set(OPENCV_DOWNLOAD_HOST "GITHUB" CACHE INTERNAL "Used for 3rdparty archives")
+# mirrors
+set(OPENCV_MIRROR_GITCODE "gitcode.net" CACHE INTERNAL "Link to mirror hosted by gitcode")
+set(OPENCV_MIRROR_CUSTOM "" CACHE INTERNAL "Link to another mirror hosted in gitlab-style")
+
+function(ocv_init_download)
+  # Run `git remote get-url origin` to get remote source
+  execute_process(
+    COMMAND
+      git remote get-url origin
+    RESULT_VARIABLE
+      RESULT_STATUS
+    OUTPUT_VARIABLE
+      OCV_GIT_ORIGIN_URL_OUT
+    ERROR_QUIET
+  )
+  # if non-git, OCV_GIT_ORIGIN_URL_OUT is empty
+  if(NOT OCV_GIT_ORIGIN_URL_OUT)
+    message(STATUS "ocv_init_download: This is not a git repo. 3rdparty resources will be downloaded from Github.com by default. You can use mirrors with CMake option -DOPENCV_MIRROR_CUSTOM=gitcode.net for example.")
+  else()
+    if(OPENCV_MIRROR_CUSTOM)
```

Comment:
> OPENCV_MIRROR_CUSTOM

If user specifies this then we should not run git check.

---

I believe variable name should be `OPENCV_DOWNLOAD_MIRROR_ID`.

Reply:
> If user specifies this then we should not run git check.

Agree. Will do.

> I believe variable name should be `OPENCV_DOWNLOAD_MIRROR_ID`.

A domain link (such as github.com, gitcode.net, etc.) should be assigned to `OPENCV_MIRROR_CUSTOM` if one wants to use customized mirror. It does not make sense to me if the CMake variable name is renamed with `*_ID`. If we are using `OPENCV_DOWNLOAD_MIRROR_ID`, the initialization process can be changed as follows:

```cmake
set(OPENCV_DOWNLOAD_MIRROR_ID "m" CACHE STRING "")

if("m${OPENCV_DOWNLOAD_MIRROR_ID}" STREUQAL "mcustom")
  set(__HOST "custom")
elseif("m${OPENCV_DOWNLOAD_MIRROR_ID}" STREUQAL "mgitcode")
  set(__HOST "gitcode")
else()
  # detect source tree via `git remote get-url origin`
endif()

if((DEFINED __HOST) AND (EXISTS "${CMAKE_SOURCE_DIR}/cmake/mirrors/${__HOST}.cmake"))
  ocv_cmake_hook_append(OCV_DOWNLOAD_MIRROR_${__HOST}  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/mirrors/${__HOST}.cmake")
else()
  # hint: not recognized host
endif()
```

In this case, domain links, download URLs, MD5s and filenames can all be organized in each `${__HOST}.cmake`.

Reply:
Old vars are all replaced with `OPENCV_DOWNLOAD_MIRROR_ID`, which should be simpler.